haystack/document_stores/elasticsearch.py

```diff
@@ -30,16 +30,31 @@
 logger = logging.getLogger(__name__)
 
 
-class ElasticsearchDocumentStore(KeywordDocumentStore):
+def prepare_hosts(host, port):
+    """
+    Create a list of host(s) + port(s) to allow direct client connections to multiple elasticsearch nodes,
+    in the format expected by the client.
+    """
+    if isinstance(host, list):
+        if isinstance(port, list):
+            if not len(port) == len(host):
+                raise ValueError("Length of list `host` must match length of list `port`")
+            hosts = [{"host": h, "port": p} for h, p in zip(host, port)]
+        else:
+            hosts = [{"host": h, "port": port} for h in host]
+    else:
+        hosts = [{"host": host, "port": port}]
+    return hosts
+
+
+class ElasticsearchCommon(KeywordDocumentStore):
```

Comment:
Two small issues here:

- I think this should be an abstract class, because it's not meant to be implemented. It has no abstract methods for now, but I imagine it will once you move to the final hierarchy that you designed in the PR description. So if this is meant to be temporary we can make this abstract later.

- I'm not too fond of the name. We use to prefix these base abstract classes with `Base`, so if I had to follow the pattern, this would be called `BaseElasticsearchDocumentStore` or anyway `Base[Something]DocumentStore`. The current `ElasticsearchCommon` is way too generic imho, doesn't imply anywhere we're even talking about a docstore. If this objection sounds too far-fetched, be aware that until very short ago we had an `ElasticsearchRetriever`, so `ElasticsearchCommon` sounds like a set of functions that could serve both, which is not correct.

Reply:
> * I think this should be an abstract class,

I tried, but in the end we don't need abstract methods here, and adding one just for the sake of make the class not instantiable is wrong IMO. I think good documentation and clean code is enough in Pythonland, but If this is important, I'd rather raise an error in the `__new__` method of the class 

> I'm not too fond of the name.

Me too, it's pretty bad and this is a code smell - a class hierarchy isn't probably the right tool here, but this is way off topic 😅. I'll change it to `BaseElasticsearchDocumentStore`.

Anyways, you touch an interesting point here:

> If this objection sounds too far-fetched, be aware that until very short ago we had an ElasticsearchRetriever, so ElasticsearchCommon sounds like a set of functions that could serve both, which is not correct.

I disagree - when thinking about names we should take into consideration the fully qualified name: `ElasticsearchCommon` alone is obviously bad, but its name is actually `haystack.document_stores.elasticsearch. ElasticsearchCommon`. It's not unusual to leverage fully qualified names in API design: for example, in Django you have a class named `DatabaseWrapper` [in every backend](https://github.com/django/django/tree/main/django/db/backends) instead of `MysqlWrapper`, `PostgresqlWrapper` and so on. Not something I would necessarily add to Haystack, just wanted to mention that packages and modules should be used to provide namespacing when writing Python APIs, as opposed to put all the effort in function or class names.

Reply:
For the abstract class, I agree that if we have to do workarounds to make it really abstract then it's worthless. +1 for keeping it as it is.

For the name, nice idea! I still think it's better to use `BaseElasticsearchDocumentStore` here and I'm glad we agree, but I'll keep it in mind for other situations. Sounds like a handy solution to keep names short.